packages/emotion-hash/hash.ml

```diff
@@ -1,133 +1,72 @@
-    let char_code_3 = Int32.of_int (Char.code str.[!index + 2]) in
-    let char_code_4 = Int32.of_int (Char.code str.[!index + 3]) in
-    (* k =
-       (str.charCodeAt(i) & 0xff) |
-       ((str.charCodeAt(++i) & 0xff) << 8) |
-       ((str.charCodeAt(++i) & 0xff) << 16) |
-       ((str.charCodeAt(++i) & 0xff) << 24) *)
-    k :=
-      char_code_1
-      ||| (char_code_2 << 8)
-      ||| (char_code_3 << 16)
-      ||| (char_code_4 << 24);
-    (* k =
-       (k & 0xffff) * 0x5bd1e995 + (((k >>> 16) * 0xe995) << 16) *)
-    k :=
-      ((!k & 0xffffl) * 0x5bd1e995l) ++ (((!k >>> 16) * 0x5bd1e995l) & 0xffffl)
-      << 16;
-    (* k ^=  k >>> 24 *)
-    k := !k ^ (!k >>> 24);
-    k :=
-      ((!k & 0xffffl) * 0x5bd1e995l) ++ (((!k >>> 16) * 0x5bd1e995l) & 0xffffl)
-      << 16;
-    (* h =
-       ((k & 0xffff) * 0x5bd1e995 + (((k >>> 16) * 0xe995) << 16)) ^
-       ((h & 0xffff) * 0x5bd1e995 + (((h >>> 16) * 0xe995) << 16)) *)
-    hash :=
-      (((!hash & 0xffffl) * 0x5bd1e995l)
-       ++ (((!hash >>> 16) * 0x5bd1e995l) & 0xffffl)
-      << 16)
-      ^ !k;
-    index := !index + 1;
+    h :=
+      add (mul (k land 0xffffl) m) (mul (k lsr 16) 0xe995l lsl 16)
+      lxor add (mul (!h land 0xffffl) m) (mul (!h lsr 16) 0xe995l lsl 16);
+
+    i := !i + 4;
     len := !len - 4
   done;
 
   (* Handle the last few bytes of the input array *)
-  (match !len with
-  | 3 ->
-    (* h ^= (str.charCodeAt(i + 2) & 0xff) << 16 *)
-    hash := !hash ^ (Int32.of_int (Char.code str.[!index + 2]) << 16)
-  | 2 ->
-    (* h ^= (str.charCodeAt(i + 1) & 0xff) << 8 *)
-    hash := !hash ^ (Int32.of_int (Char.code str.[!index + 1]) << 8)
-  | 1 ->
-    (* h ^= str.charCodeAt(i) & 0xff *)
-    hash := !hash ^ Int32.of_int (Char.code str.[!index]);
-    (* h =
-       (h & 0xffff) * 0x5bd1e995 + (((h >>> 16) * 0xe995) << 16) *)
-    hash :=
-      ((!hash & 0xffffl) * 0x5bd1e995l)
-      ++ (((!hash >>> 16) * 0x5bd1e995l) & 0xffffl)
-      << 16
-  | _ -> ());
-
-  (* Do a few final mixes of the hash to ensure the last few bytes are well-incorporated. *)
+  let () =
+    if !len = 3 then h := !h lxor ((of_char str.[!i + 2] land 0xffl) lsl 16);
+    if !len >= 2 then begin
+      h := !h lxor ((of_char str.[!i + 1] land 0xffl) lsl 8)
+    end;
+    if !len >= 1 then begin
+      h := !h lxor (of_char str.[!i] land 0xffl);
+      h := add (mul (!h land 0xffffl) m) (mul (!h lsr 16) 0xe995l lsl 16)
+    end
+  in
 
-  (* h ^= h >>> 13 *)
-  hash := !hash ^ (!hash >>> 13);
-  (* h =
-     (h & 0xffff) * 0x5bd1e995 + (((h >>> 16) * 0xe995) << 16) *)
-  hash :=
-    ((!hash & 0xffffl) * 0x5bd1e995l) ++ ((!hash >>> 16) * 0x5bd1e995l << 16);
-  (* (h ^ (h >>> 15)) >>> 0 *)
-  !hash ^ (!hash >>> 15) >>> 0
+  (* Do a few final mixes of the hash to ensure the last few
+     bytes are well-incorporated. *)
+  h := !h lxor (!h lsr 13);
 
-let default (str : string) = str |> murmur2 |> to_css
+  h := add (mul (!h land 0xffffl) m) (mul (!h lsr 16) 0xe995l lsl 16);
```

Comment:
idk how to override the `+` and `*` to use the Int32 ones. That'd make this code a bit easier to read.